### PR TITLE
Add abstract `ErrorFlusherBase` class

### PR DESCRIPTION
**Pull request by vilhena-stripe**
Renames `ErrorFlusher` to `ErrorFlusherStdout`

### Motivation
First step of final milestone of the sorbet fast feedback project. We'll eventually have multiple error flusher classes. This does not change any existing functionality.

### CODE REVIEW

File: core/ErrorFlusherBase.h [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+#include "core/ErrorQueueMessage.h"
+#include <vector>
+
+namespace sorbet {
+class FileSystem;
+namespace core {
+
+class ErrorFlusherBase {
+public:
+    virtual void flushErrors(spdlog::logger &logger, std::vector<std::unique_ptr<ErrorQueueMessage>> error) = 0;
```
Comment (DarkDimius) [inline core/ErrorFlusherBase.h:14]: You need a virtual destructor

File: core/ErrorQueue.h [hunk 2]
```diff
@@ -17,7 +17,7 @@ class ErrorQueue {
     std::vector<std::unique_ptr<ErrorQueueMessage>> drainAll();
     std::vector<std::unique_ptr<ErrorQueueMessage>> drainFlushed();
     void collectForFile(core::FileRef whatFile, std::vector<std::unique_ptr<core::ErrorQueueMessage>> &out);
-    ErrorFlusher errorFlusher;
+    ErrorFlusherStdout errorFlusher;
```
Comment (jvilk-stripe) [inline core/ErrorQueue.h:20]: This should be an `ErrorFlusher` rather than the specific `ErrorFlusherStdout`. You can then instantiate it to an `ErrorFlusherStdout` instance in the `ErrorQueue` constructor. Later, we'll pass that responsibility to the callsite creating the error queue.

(You can make this a `unique_ptr` to make that work.)

**Comment by DarkDimius** [comment]
Are there some schemes that would allow us not to use multiple classes and have more uniform behaviour? 

Also, if we're starting to use virtual methods here, new pitfalls of C++ can appear. If you use virtual methods you almost always want to store  objects by pointer rather than by value, as otherwise surprises await you. But this is secondary to "can we do this more uniformly and thus have a single class"?

**Comment by jvilk-stripe** [comment]
@DarkDimius I'm not sure how to do that.

CLI wants errors to be streamed to stderr, LSP wants errors to be streamed to client, queries want errors to be suppressed...

We could pass in... another object that error flusher flushes to? A lambda? But that seems almost as bad.

**Comment by jvilk-stripe** [comment]
At its core, `ErrorFlusher` is just hooking up flushed `ErrorQueue` messages to some output device. There's not much extraneous duplicated logic there. It seemed natural to use classes for this, but I'm open to other ideas.

File: core/ErrorFlusherBase.h [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+#include "core/AutocorrectSuggestion.h"
+#include "core/ErrorQueueMessage.h"
+#include <vector>
+
+namespace sorbet {
+class FileSystem;
+namespace core {
+
+class ErrorFlusherBase {
```
[the remark this answers is not in the document]
Comment (vilhena-stripe) [reply core/ErrorFlusherBase.h:12]: Sure, wasn't sure of the naming convention

File: core/ErrorFlusher.h [hunk 4]
```diff
@@ -10,15 +10,11 @@ class FileSystem;
...
-
 public:
-    ErrorFlusher() = default;
-    void flushErrors(spdlog::logger &logger, std::vector<std::unique_ptr<ErrorQueueMessage>> error);
-    void flushErrorCount(spdlog::logger &logger, int count);
-    void flushAutocorrects(const GlobalState &gs, FileSystem &fs);
+    virtual void flushErrors(spdlog::logger &logger, std::vector<std::unique_ptr<ErrorQueueMessage>> error) = 0;
+    virtual void flushErrorCount(spdlog::logger &logger, int count) = 0;
+    virtual void flushAutocorrects(const GlobalState &gs, FileSystem &fs) = 0;
+    virtual ~ErrorFlusher(){};
```
Comment (jvilk-stripe) [inline core/ErrorFlusher.h:17]: You can just make this the default.
```suggestion
    virtual ~ErrorFlusher() = default;
```

File: core/ErrorQueue.h (hunk 2 above)
Comment (vilhena-stripe) [reply core/ErrorQueue.h]: Why does it have to be `unique_ptr` to work?

File: core/ErrorFlusherStdout.h [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+namespace sorbet::core {
+
+class ErrorFlusherStdout : public ErrorFlusher {
+private:
+    std::vector<AutocorrectSuggestion> autocorrects;
+    bool printedAtLeastOneError{false};
+
+public:
+    ErrorFlusherStdout() = default;
+    ~ErrorFlusherStdout(){};
```
Comment (jvilk-stripe) [inline core/ErrorFlusherStdout.h:18]:
```suggestion
    ~ErrorFlusherStdout() = default;
```

File: core/ErrorFlusherStdout.h [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    using ErrorFlusher::flushErrors;
```
Comment (jvilk-stripe) [inline core/ErrorFlusherStdout.h:19]: Why are these here? You shouldn't need these.

File: core/ErrorFlusherStdout.h [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    virtual void flushErrors(spdlog::logger &logger, std::vector<std::unique_ptr<ErrorQueueMessage>> error);
```
Comment (jvilk-stripe) [inline core/ErrorFlusherStdout.h:20]: This and the others should be `override`

```suggestion
    void flushErrors(spdlog::logger &logger, std::vector<std::unique_ptr<ErrorQueueMessage>> error) override;
```

File: core/ErrorQueue.h [hunk 8]
```diff
@@ -17,7 +17,7 @@ class ErrorQueue {
     std::vector<std::unique_ptr<ErrorQueueMessage>> drainAll();
     std::vector<std::unique_ptr<ErrorQueueMessage>> drainFlushed();
     void collectForFile(core::FileRef whatFile, std::vector<std::unique_ptr<core::ErrorQueueMessage>> &out);
-    ErrorFlusher errorFlusher;
+    std::unique_ptr<ErrorFlusher> errorFlusher;
```
Comment (jvilk-stripe) [inline core/ErrorQueue.h:20]: Let's actually make this a `shared_ptr`; we don't need to ensure unique ownership here, and it'll make downcasting easier if we ever need to.

File: core/ErrorFlusherStdout.h (hunk 6 above)
Comment (vilhena-stripe) [reply core/ErrorFlusherStdout.h:19]: I was getting a warning that I was hiding overloaded virtual function, and tried to find the answer online, and that's what I found

File: core/ErrorQueue.h (hunk 2 above)
Comment (jvilk-stripe) [reply core/ErrorQueue.h]: It has to be a pointer or reference type to work. The reason why has to do with determining the size of objects and storing things by value.

Most programming languages only support referring to objects by reference. For example, in Java, having a field like so:

```
class Blah {
ErrorFlusher errorFlusher;
}
```

...means that objects of type `Blah` have a field `errorFlusher` containing a _reference_ to an object with type `ErrorFlusher`. The size of that reference is constant -- it's the size of a pointer. Thus, we can store a reference to `ErrorFlusher` or any of its subclasses in the same amount of space.

In C++, this syntax has a different meaning. `ErrorFlusher errorFlusher` says that the object _contains_ a full `ErrorFlusher` object. In other words, the size of Blah in memory is roughly sizeof(ErrorFlusher) + sizeof(other fields on Blah).

Subclasses of `ErrorFlusher` may take more space than the base `ErrorFlusher` class, and, for that reason, cannot be stored by value as `ErrorFlusher`. Thus, we actually want to store a (smart) pointer to `ErrorFlusher` instead, which will let us store any subclass of `ErrorFlusher` in the field in the same amount of space.

There are other good reasons to use pointers over raw value types:

* **Reduces copying.** Assigning a new ErrorFlusher means that we are copying the contents of the first ErrorFlusher. Copying can be super cheap if ErrorFlusher only contains simple data types, but if it contains any object with a custom copy constructor then that code will have to run to copy that object.
  * tl;dr: We reserve referring to objects by value for super simple data types (like SymbolRef).
  * Fun note: Yeah, you can tell C++ how to copy your object.
* **Lets you forward declare to reduce #imports in headers.** Header files only need to know the full type information for types that are 1) stored by value in order to figure out the size of things in memory and 2) that are actually accessed in method implementations or field initialization in the header. References and pointer types are always the same size, so you can just tell the compiler with a forward decl that they exist and shave off #imports  (which can make compilation faster).
